Plumb timeout through ReadAndDispatch, Shell(), use in long-running cmds

erock2112 opened the pull request:
This fixes a bug in which ReadAndDispatch hits a timeout for a
long-running command (eg. 'pm list packages') and assumes that the
connection was broken. When the command finally finishes, we get errors
when reading output from subsequent commands.

maruel commented on adb/contrib/high.py:
```diff
@@ -729,7 +729,7 @@ def GetLastUID(self):
   def GetPackages(self):
     """Returns the list of packages installed."""
-    out, _ = self.Shell('pm list packages')
+    out, _ = self.Shell('pm list packages', timeout_ms=30000)
```
Please use min() when self._timeout is not none.

erock2112 replied:
HighDevice does not store a default timeout.  Added the max(timeout, default) logic to AdbCommandsSafe and to common.Handle.

maruel replied:
Please add a short comment as to why, e.g. "pm can be very slow at time, and a short default timeout can be confused with the device being unresponsive when in practice it's just pm that took a lot of time to return."

erock2112 replied:
Done.

maruel commented on adb/contrib/high.py:
```diff
@@ -818,7 +818,7 @@ def WaitUntilFullyBooted(self, timeout=300):
             self.port_path)
         return False
-      out, _ = self.Shell('pm path')
+      out, _ = self.Shell('pm path', timeout_ms=30000)
```
sane

maruel replied:
lol I hate github

erock2112 commented on adb/adb_protocol.py:
```diff
@@ -175,7 +175,6 @@ def __init__(self, header, data=''):
 
   def Write(self, usb, timeout_ms=None):
     """Send this message over USB."""
-    timeout_ms = usb.Timeout(timeout_ms)
```
Removed call to usb.Timeout() here and in Read() because usb.BulkRead and usb.BulkWrite both already use self.Timeout() internally.

erock2112 commented on adb/contrib/adb_commands_safe.py:
```diff
@@ -433,8 +433,11 @@ def ShellRaw(self, cmd):
-    out = self._adb_cmd.Shell(cmd + self._SHELL_SUFFIX).decode(
-        'utf-8', 'replace')
+    timeout_ms = max(
+        timeout_ms or self._default_timeout_ms, self._default_timeout_ms)
```
This might actually not be necessary, but I'm not totally clear on all of the ways that the USB handle can be created.  If it has the same default timeout as the AdbCommandsSafe object, then this logic is unnecessary.

maruel replied:
@bpastene about other use case.

maruel commented on adb/common.py:
```diff
@@ -63,7 +63,7 @@ def __init__(self, serial=None, timeout_ms=None):
 
   def Timeout(self, timeout_ms):
-    return timeout_ms if timeout_ms is not None else self._timeout_ms
+    return max(timeout_ms or self._timeout_ms, self._timeout_ms)
```
> > > max(None, 10)
> > > 10
> > > max(10, None)
> > > 10

erock2112 replied:
Done.

maruel commented on adb/contrib/adb_commands_safe.py:
```diff
@@ -433,8 +433,11 @@ def ShellRaw(self, cmd):
     assert self.IsShellOk(cmd), 'Command is too long: %r' % cmd
-    out = self._adb_cmd.Shell(cmd + self._SHELL_SUFFIX).decode(
-        'utf-8', 'replace')
+    timeout_ms = max(
```
Same

maruel commented:
Can you squash your commits?

erock2112 committed:
Plumb timeout through ReadAndDispatch, Shell(), use in long-running cmds

This fixes a bug in which ReadAndDispatch hits a timeout for a
long-running command (eg. 'pm list packages') and assumes that the
connection was broken. When the command finally finishes, we get errors
when reading output from subsequent commands.

erock2112 commented:
Squashed.

maruel commented:
Thanks!
